# DSCI 100 Project: Final Report

**Authors**: Rodrigo Moreno González, Trevor Lau, Sydney Peters, Megan Zhang


**Predicting Usage of a Video Game Research Server**
A research group in the Department of Computer Science at UBC (PLAI) has a goal to aid in advancing the limits of artificial intelligence. They host a Minecraft server that records players' actions as they navigate through the world. Since running the project is so complex, they want to recruit players who will contribute large amounts of gameplay data, as well as make sure they have the resources to handle the amount of players they attract. In an attempt to optimize their players' contributions, they want to understand what demographics of players will have the greatest gameplay time. This project answers their question of how player demographics relate to total gameplay time. 

## The Question
**Question 2**: Which kinds of players are most likely to contribute a large amount of data, and how can we identify them for targeted recruitment?

**Predictive Question** Can we predict the total `played_hours` from the `players.csv` dataset based on the predictor, `experience`. 

**Response variable**:

`played_hours` - total play time and data contribution

**Explanatory variable**:

`experience` - player experience level

The dataset allows us to examine if `experience` relates to total playtime. If certain experience levels tend to play more, those groups may contribute more gameplay data. This helps identify the player demographics most likely to provide substantial data for the research team.


## Description: The Players
The dataset used in this analysis is a player demographics and engagement dataset that contains 196 observations (rows) and 9 variables (columns). The dataset records information about individual players, their gaming experience level, subscription status, demographic details, and the amount of time they have spent playing.

- Number of observations: 196 (each row represents one unique player)
- Number of variables: 9
- Observational unit: Player-level data
- Purpose: Identify which demographic characteristics correspond to higher gameplay time (and therefore more data contributed)
  
**Variables**
- `played_hours` (numeric) - Total hours each player spent on the server
- `experience` (categorical) -  Self-reported experience level in Minecraft
- `gender` (categorical) - Player’s gender
- `subscribe` (Boolean) - Player’s subscription status
- `hashedEmail` (String) - Player’s email address, hashed for privacy protection
- `name` (String) - Player’s name
- `age` (Numeric) - Player’s age
- `individualID` (String) - Unique player identifier
- `organizationName` (String) - Name of the organization associated with the player

**Issues to consider**
- Missing identifiers (`individualId`, `organizationName`).
- `hashedEmail` is anonymous for privacy.
- `played_hours` may include idle time, which could overestimate actual engagement.
- `experience` is self reported which could lead to inconsistent standards. 

Source of the original dataset:
https://drive.google.com/file/d/1Mw9vW0hjTJwRWx0bDXiSpYsO3gKogaPz/edit


In [1]:
# Import libraries
import pandas as pd
import altair as alt

# Import the K-NN regression model
from sklearn.model_selection import GridSearchCV, cross_validate, train_test_split
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn import set_config

# Import the K-NN regression model
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error

# Read the datasets
url= "https://raw.githubusercontent.com/sydlpeters/dsci-group-2025w1-group-101-1/refs/heads/main/data/players.csv"
# Load the dataset 
players = pd.read_csv(url)
players

experience  subscribe                                        hashedEmail  \
0          Pro       True  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1      Veteran       True  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2      Veteran      False  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3      Amateur       True  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4      Regular       True  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   
..         ...        ...                                                ...   
191    Amateur       True  b6e9e593b9ec51c5e335457341c324c34a2239531e1890...   
192    Veteran      False  71453e425f07d10da4fa2b349c83e73ccdf0fb3312f778...   
193    Amateur      False  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
194    Amateur      False  f19e136ddde68f365afc860c725ccff54307dedd13968e...   
195        Pro       True  d9473710057f7d42f36570f0be83817a4eea614029ff90...   

     played_hours       name             gender  age  individualId  \
0            30.3     Morgan               Male    9           NaN   
1             3.8  Christian               Male   17           NaN   
2             0.0      Blake               Male   17           NaN   
3             0.7      Flora             Female   21           NaN   
4             0.1      Kylie               Male   21           NaN   
..            ...        ...                ...  ...           ...   
191           0.0     Bailey             Female   17           NaN   
192           0.3     Pascal               Male   22           NaN   
193           0.0      Dylan  Prefer not to say   17           NaN   
194           2.3     Harlow               Male   17           NaN   
195           0.2      Ahmed              Other   91           NaN   

     organizationName  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
..                ...  
191               NaN  
192               NaN  
193               NaN  
194               NaN  
195               NaN  

[196 rows x 9 columns]

In [2]:
# Experience comapred with total played hours played
plot1= alt.Chart(players, title="Fig 1. Total Hours Played per Experience Group").mark_bar().encode(
    x=alt.X("experience:N").title("Experience Level"),
    y=alt.Y("played_hours:Q").title("Average hours Played(hr)"),
    color = alt.Color("experience:N").title("Experience Level")
)
plot1

alt.Chart(...)

In [3]:
# Experience comapred with average played hours played
plot2= alt.Chart(players, title="Fig. 2 Average Hours Played per Experience Group").mark_bar().encode(
    x=alt.X("experience:N").title("Experience Level"),
    y=alt.Y("mean(played_hours):Q").title("Average hours Played(hr)"),
    color = alt.Color("experience:N").title("Experience Level")
)
plot2

alt.Chart(...)

In [4]:
# Subscription status comapred with average hours played
plot3= alt.Chart(players, title="Fig. 3 Average Hours by Subscribed/Unsubscribed").mark_bar().encode(
    x=alt.X("subscribe:N").title("Subscription Status"),
    y=alt.Y("mean(played_hours):Q").title("Average Hours Played(hr)"),
    color = alt.Color("subscribe:N").title("Subscription Status")
)
plot3

alt.Chart(...)

In [5]:
# Gender comapred with average hours played
plot4= alt.Chart(players, title="Fig. 4 Average Hours Played by Gender").mark_bar().encode(
    x=alt.X("gender:N").title("Gender"),
    y=alt.Y("mean(played_hours):Q").title("Average Hours Played(hr)"),
    color = alt.Color("gender:N").title("Gender")
)
plot4

alt.Chart(...)

In [6]:
# Age comapred with average hours played 
plot5 = alt.Chart(players, title="Fig 5. Average Hours Played by Age").mark_circle(opacity=0.8).encode(
    x=alt.X("age:Q").title("Age"),
    y=alt.Y("mean(played_hours):Q").title("Average Hours Played (hr)"),
    color=alt.Color("age:Q").title("Age")
)
plot5

alt.Chart(...)

**Visualizations Summary**

Across all visualizations, the data show that total gameplay hours are heavily right-skewed, with most players recording very little playtime and a small number accounting for the largest values.

Experience level (Figures 1 & 2) displays the clearest relationship with gameplay time: average hours generally increase with higher experience categories, suggesting a positive association between self-reported experience and total playtime. Although variability is high within each group, the upward trend indicates that experience is meaningfully related to engagement.

Subscription status (Figure 3) shows some difference between subscribed and unsubscribed players, with subscribed players tending to average slightly more gameplay hours. However, the distributions overlap substantially, meaning subscription alone does not strongly separate players by playtime.

Gender (Figure 4) reveals no consistent pattern across groups; average hours between genders appear similar, with any observed differences likely driven by small sample sizes or extreme outliers rather than true group-level effects. This suggests gender provides little predictive value for modeling gameplay time.

Age (Figure 5) shows no strong monotonic trend. A few ages exhibit higher average values, but these spikes are inconsistent and likely reflect sparse data and outliers rather than a genuine relationship. Overall, age does not demonstrate a stable or reliable association with total played hours

Self-reported experience is the best predictor for KNN regression in this dataset, as it exhibits the strongest and most consistent relationship to total gameplay time. Subscription status may provide minor additional information, while age and gender contribute little meaningful predictive value and should not be heavily weighted in the model.


In [7]:
# Drop unused and non-predictive columns
players_tidy = players.drop(columns=["individualId", "organizationName", "name", "gender", "hashedEmail", "age", "subscribe"])
# Preview tidy players_tidy
players_tidy

experience  played_hours
0          Pro          30.3
1      Veteran           3.8
2      Veteran           0.0
3      Amateur           0.7
4      Regular           0.1
..         ...           ...
191    Amateur           0.0
192    Veteran           0.3
193    Amateur           0.0
194    Amateur           2.3
195        Pro           0.2

[196 rows x 2 columns]

In [8]:
# View full summary (numeric + categorical)
players_tidy.describe(include="all")

experience  played_hours
count         196    196.000000
unique          5           NaN
top       Amateur           NaN
freq           63           NaN
mean          NaN      5.845918
std           NaN     28.357343
min           NaN      0.000000
25%           NaN      0.000000
50%           NaN      0.100000
75%           NaN      0.600000
max           NaN    223.100000

**Statistics Summary**

Most players recorded very little playtime, with three-quarters spending under one hour in total (Q3 = 0.6 hours). Only a small number logged substantially more time, which raises the overall average (mean = 5.85 hours; median = 0.1 hours) and contributes to a large spread in the data (SD = 28.36 hours). This indicates that most players played very little (close to 0 hours), while a few played a great deal, making the distribution highly uneven and right-skewed. The experience variable includes several levels that are not evenly represented, with some experience groups appearing more frequently than others. Overall, the dataset shows substantial variability in both gameplay time and experience level, supporting the use of KNN regression to explore patterns between player experience and total playtime.

In [9]:
# Create a numeric mapping for ordered experience levels
experience_map = {
    "Beginner" : 0,
    "Regular": 1,
    "Amateur": 2,
    "Veteran": 3,
    "Pro": 4
}
# Convert the categorical experience column to numeric values
players_tidy["experience"] = players_tidy["experience"].map(experience_map)
players_tidy

experience  played_hours
0             4          30.3
1             3           3.8
2             3           0.0
3             2           0.7
4             1           0.1
..          ...           ...
191           2           0.0
192           3           0.3
193           2           0.0
194           2           2.3
195           4           0.2

[196 rows x 2 columns]

In [10]:
# Split the data
players_training, players_testing = train_test_split(
    players_tidy, 
    test_size=0.2, 
    random_state=1234
)

# Set target and predictors
X_train = players_training[["experience"]]
y_train = players_training["played_hours"]

X_test = players_testing[["experience"]]
y_test = players_testing["played_hours"]

In [11]:
# Create preprocessor
players_preprocessor = make_column_transformer(
    (StandardScaler(), ["experience"]),
    remainder="drop",
    verbose_feature_names_out=False,
)

# Create pipeline
players_pipeline = make_pipeline(
    players_preprocessor,
    KNeighborsRegressor()
)

# Evaluate the performance of players_pipeline using 5-fold cross-validation
players_cv = pd.DataFrame(
    cross_validate(
        players_pipeline,
        X_train,
        y_train,
        cv=5,
        scoring="neg_root_mean_squared_error",
        return_train_score=True,
    )
)

players_cv

fit_time  score_time  test_score  train_score
0  0.004293    0.002584  -19.160247   -27.106263
1  0.003353    0.002203   -5.803200   -21.794819
2  0.003131    0.001952  -29.223031   -16.374080
3  0.003076    0.001908   -3.798582   -21.907992
4  0.003044    0.001887  -31.885128   -15.327369

In [12]:
# Create the 5-fold GridSearchCV object
param_grid = {
    "kneighborsregressor__n_neighbors": range(1, 51),
}

players_tuned = GridSearchCV(
    estimator=players_pipeline,
    param_grid=param_grid,
    cv=5,
    scoring="neg_root_mean_squared_error",
)

# fit the GridSearchCV object
players_result = pd.DataFrame(players_tuned.fit(X_train, y_train).cv_results_)

players_result

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.003446      0.000330         0.002164        0.000078   
1        0.003052      0.000034         0.018451        0.033085   
2        0.003101      0.000125         0.001915        0.000037   
3        0.003057      0.000067         0.004010        0.004209   
4        0.003049      0.000060         0.001905        0.000018   
5        0.002994      0.000011         0.001898        0.000034   
6        0.003019      0.000052         0.003396        0.003047   
7        0.003003      0.000052         0.001876        0.000005   
8        0.003002      0.000054         0.001868        0.000012   
9        0.002970      0.000006         0.001873        0.000006   
10       0.003009      0.000056         0.001872        0.000009   
11       0.003012      0.000054         0.001879        0.000004   
12       0.002965      0.000035         0.001866        0.000010   
13       0.002970      0.000063         0.001853        0.000003   
14       0.002966      0.000053         0.001912        0.000115   
15       0.002962      0.000054         0.001856        0.000005   
16       0.002944      0.000014         0.001886        0.000059   
17       0.002972      0.000051         0.001858        0.000010   
18       0.002976      0.000076         0.001857        0.000009   
19       0.002944      0.000003         0.001884        0.000016   
20       0.003008      0.000121         0.001866        0.000007   
21       0.002968      0.000057         0.001864        0.000014   
22       0.002968      0.000065         0.001867        0.000007   
23       0.003111      0.000238         0.002195        0.000509   
24       0.005836      0.005431         0.001940        0.000025   
25       0.003013      0.000059         0.001930        0.000061   
26       0.002985      0.000010         0.001925        0.000022   
27       0.003029      0.000076         0.001921        0.000023   
28       0.003103      0.000219         0.001907        0.000013   
29       0.002989      0.000064         0.001896        0.000004   
30       0.002958      0.000003         0.001895        0.000005   
31       0.002988      0.000056         0.001893        0.000008   
32       0.002989      0.000054         0.001895        0.000007   
33       0.003130      0.000338         0.001905        0.000007   
34       0.002988      0.000065         0.001895        0.000002   
35       0.002995      0.000076         0.001892        0.000010   
36       0.002986      0.000051         0.001895        0.000010   
37       0.002952      0.000005         0.001916        0.000035   
38       0.002982      0.000053         0.001925        0.000060   
39       0.002980      0.000053         0.001896        0.000003   
40       0.002952      0.000007         0.001935        0.000060   
41       0.002975      0.000056         0.001901        0.000006   
42       0.002977      0.000053         0.001904        0.000004   
43       0.002946      0.000010         0.001909        0.000013   
44       0.002997      0.000076         0.001913        0.000007   
45       0.002994      0.000058         0.001919        0.000005   
46       0.002955      0.000007         0.001917        0.000001   
47       0.002987      0.000060         0.001920        0.000008   
48       0.002947      0.000008         0.001931        0.000036   
49       0.002980      0.000057         0.001926        0.000013   

    param_kneighborsregressor__n_neighbors  \
0                                        1   
1                                        2   
2                                        3   
3                                        4   
4                                        5   
5                                        6   
6                                        7   
7                                        8   
8                                        9   
9                                       10   
10                           

In [13]:
# Get the best parameter values
players_min = players_tuned.best_params_
players_min

{'kneighborsregressor__n_neighbors': 39}

In [14]:
# Extract the best (lowest) cross-validation error from the tuned KNN model
players_best_RMSPE = -players_tuned.best_score_
players_best_RMSPE

np.float64(16.028531937252072)

**Cross-Validation Summary**

Grid search cross-validation identified the optimal number of neighbors for the KNN regression model as K = 39. This means the model achieved its best predictive performance when each prediction was calculated by averaging the outcomes of the 39 most similar players. Using a relatively large value of K helps smooth predictions and reduce the influence of extreme outliers in the dataset, which is particularly beneficial given the highly skewed distribution of gameplay hours. 

The tuned KNN regression model achieved a cross-validated Root Mean Squared Prediction Error (RMSPE) of approximately 16.03 hours. This means that, on average, the model’s predicted gameplay time differs from the true recorded value by about 16 hours per player. Given the highly right-skewed distribution of playtime—where most players logged very little activity and a small number recorded extremely high usage—this level of error is expected and highlights the difficulty of accurately predicting individual play behavior. Overall, the model captures general trends related to player experience but remains limited in precision when forecasting exact playtime for each player.

In [15]:
# Set K, make pipeline, and fit the training data
knn = KNeighborsRegressor(n_neighbors = 39)
player_pipe = make_pipeline(players_preprocessor, knn)
player_pipe.fit(X_train, y_train)

# Predicte total hours based on experience level from testing set 
players_predictions = players_testing.assign(prediction = player_pipe.predict(X_test))
players_predictions

experience  played_hours  prediction
101           2           0.0    7.261538
51            1         218.1    6.494872
146           4           0.0    1.305128
153           0           0.1    1.410256
106           1           0.0    6.494872
59            3           0.2    0.464103
161           3           0.0    0.464103
167           0           0.3    1.410256
193           2           0.0    7.261538
88            0           0.0    1.410256
29            3           0.1    0.464103
130           2          56.1    7.261538
66            3           0.1    0.464103
74            1         223.1    6.494872
48            3          12.5    0.464103
100           2           0.0    7.261538
123           0           7.1    1.410256
31            3           0.1    0.464103
95            1           0.0    6.494872
137           3           0.0    0.464103
6             1           0.0    6.494872
54            0           0.0    1.410256
97            3           0.1    0.464103
5             2           0.0    7.261538
110           2           0.0    7.261538
91            0           0.0    1.410256
134           0           1.1    1.410256
180           2           2.1    7.261538
7             2           0.0    7.261538
168           3           0.1    0.464103
94            0           0.8    1.410256
9             3           0.0    0.464103
58            1           3.6    6.494872
92            4           1.2    1.305128
111           1           4.0    6.494872
43            3           0.0    0.464103
77            1           0.0    6.494872
142           0           1.0    1.410256
119           0           0.0    1.410256
40            1           5.6    6.494872

In [16]:
# Predicted vs total played hours by experience 
players_plot = (alt.Chart(players_predictions).mark_circle(opacity=0.7).encode(
    x=alt.X("experience").title("Experience").scale(zero=False),
    y=alt.Y("played_hours").title("played_hours").scale(zero=False)
)+
alt.Chart(
    players_predictions,
    title= "Fig. 6 Predicted vs Total Gameplay Hours by Experience (K = 39)"
).mark_line(
    color="black"
).encode(
    x="experience",
    y="prediction"
))
players_plot

alt.LayerChart(...)

**KNN Model and Visualization Analysis**

The KNN regression model was constructed using the optimal hyperparameter value of K = 39 neighbors, as determined through cross-validated grid search tuning. A preprocessing pipeline was applied to standardize the predictor variable (experience) prior to modeling, which is essential for distance-based algorithms such as KNN to ensure valid similarity calculations. The model was trained on 80% of the cleaned dataset and then used to generate predictions for the remaining 20% held-out test set. Predicted playtime values were calculated as the average observed gameplay hours of the 39 most similar players in the training set based on experience level and were appended to the testing dataset for evaluation and visualization.

The scatterplot reveals a right-skewed distribution of playtime, with most players exhibiting very low gameplay hours and a small number recording extremely high usage. This results in wide vertical dispersion at all experience levels and indicates substantial individual variability that limits precise prediction accuracy. Despite this variability, the prediction line displays a clear positive upward trend, showing that regular and amateur experience levels are associated with increasing average gameplay hours. This supports the earlier conclusion that experience is the strongest predictor variable in the dataset. The smooth shape of the fitted prediction line reflects the use of a relatively large neighborhood size (K = 39). This causes each prediction to be lightly smoothed across many neighboring data points, reducing overfitting to extreme values and generating more stable trend estimates. Consequently, the model tends to predict conservative values:

- Under-predicting extreme high-usage players, and
- Slightly over-predicting very low-usage players.

This averaging behavior improves generalization but limits individual-level prediction precision, which aligns with the validated performance metrics.

In [17]:
# Predicte total hours based on experience level from whole data set
players_all_predictions = players_tidy.assign(
    prediction_all=player_pipe.predict(
        players_tidy[["experience"]]
    )
)
players_all_predictions

experience  played_hours  prediction_all
0             4          30.3        1.305128
1             3           3.8        0.464103
2             3           0.0        0.464103
3             2           0.7        7.261538
4             1           0.1        6.494872
..          ...           ...             ...
191           2           0.0        7.261538
192           3           0.3        0.464103
193           2           0.0        7.261538
194           2           2.3        7.261538
195           4           0.2        1.305128

[196 rows x 3 columns]

In [19]:
# All predicted vs total played hours by experience 
players_plot = (alt.Chart(players_all_predictions).mark_circle(opacity=0.7).encode(
    x=alt.X("experience").title("Experience").scale(zero=False),
    y=alt.Y("played_hours").title("played_hours").scale(zero=False)
)+
alt.Chart(
    players_all_predictions,
    title= "Fig. 7 All Predicted vs. Actual Gameplay Hours by Experience (K = 39)"
).mark_line(
    color="black"
).encode(
    x="experience",
    y="prediction_all"
))
players_plot

alt.LayerChart(...)

**KNN Model and Visualization Analysis - Prediction All**

The model tends to underestimate the most extreme high-playtime outliers, and slightly overestimate some of the very low-playtime players. This behavior is expected for a KNN model with a large K: predictions are essentially local averages, so they pull extreme values back toward typical levels. When plotted across the full dataset, this makes the line a good representation of the average engagement pattern rather than individual outcomes.

The scatter layer using all players reinforces earlier findings: The distribution of played_hours is highly right-skewed, meaning that most players record very low total hours, while a relatively small number log extremely high values. There is substantial vertical spread at each experience level, highlighting large individual differences in engagement even among players with similar experience. Despite this noise, there is a clear overall positive relationship between experience and gameplay hours, with regular and amateur players tending to have higher total playtime on average. Because all observations are included, clusters and density at low playtime values become more evident, emphasizing just how concentrated the majority of players are near zero hours.

## Discussion 
**Summary of Findings**

This analysis found that player experience level is a strong predictor of total gameplay time, with notable differences between experience categories. While higher experience generally corresponded with increased engagement, the Amateur experience group (2) had the highest average playtime among all categories, exceeding even Veteran and Pro players. This suggests that engagement does not increase strictly monotonically with experience level; instead, players in the Amateur stage may represent a peak period of gameplay activity. Across all groups, gameplay hours were highly right-skewed, with most players recording minimal playtime and a small number logging extremely high values. This variability contributed to a cross-validated Root Mean Squared Prediction Error of approximately 16.03 hours, even for the best-performing KNN model (K = 39), indicating that individual playtime remains difficult to predict precisely.

**Comparison to Expectations**

This result was partially expected but also somewhat surprising. It was anticipated that engagement would increase with rising experience; however, the finding that Amateur players exhibited the greatest playtime suggests that engagement may peak during the middle stages of skill development rather than at the highest levels. This aligns with behavior seen in many gaming communities, where players are most active during skill-building phases and may reduce playtime once they reach advanced mastery or specific personal goals. 

**Potential Impacts of These Findings**

The identification of Amateur players as the most engaged group has several practical implications:
- Retention strategies could focus on supporting players during the transition from Regular to Amateur to sustain peak engagement.
- Game content and challenges could be tailored to target Amateur-level skill development, where players may be most motivated to improve.
- Recruitment for gameplay data collection could prioritize Amateur players, as they appear likely to generate the largest gameplay volumes.
- Predictive models can use experience as a key feature while acknowledging that peak engagement may occur at intermediate rather than maximum experience levels.

**Future Directions**

These findings lead to several future research questions:
- Why does engagement peak at the Amateur level rather than at Veteran or Pro levels?
- Are Amateur players experiencing higher motivation, progression goals, or community involvement compared to other groups?
- Would adding measures such as session frequency, achievement completion, or social gameplay features better explain this pattern?
- Does the engagement peak persist across time, or does it differ for newer player cohorts?

Exploring these questions could deepen understanding of player engagement cycles and further enhance predictive modeling accuracy.

## Works Cited
Pacific Laboratory for Artificial Intelligence (PLAI). “Blocked.” Plaicraft.ai, 2025, www.plaicraft.ai. Accessed 3 Dec. 2025.